src/discovery.rs

```diff
@@ -34,34 +39,79 @@ const QUERY: &str = r#"{
-/// a problem decoding the response.
-pub fn with_timeout(timeout: Option<Duration>) -> Result<Vec<(SocketAddr, DeviceData)>> {
-    let socket = UdpSocket::bind("0.0.0.0:0")?;
-    socket.set_broadcast(true)?;
-    socket.set_read_timeout(timeout)?;
-
-    let req = protocol::encrypt(QUERY)?;
+fn can_interface_broadcast(iface: Interface) -> Option<(Ipv4Addr, Ipv4Addr)> {
+    match iface.addr {
+        IfAddr::V4(addr) => match (addr.ip.is_loopback(), addr.broadcast) {
+            (false, Some(broadcast)) => Some((addr.ip, broadcast)),
+            _ => None,
+        },
+        IfAddr::V6(_) => None,
+    }
+}
 
+fn discover_on_interface(
+    timeout: Option<Duration>,
+    ip: Ipv4Addr,
+    broadcast: Ipv4Addr,
+    request: &[u8],
+) -> Result<HashMap<SocketAddr, DeviceData>> {
+    let socket_addr = SocketAddr::new(IpAddr::V4(ip), 0);
+    let udp_socket = UdpSocket::bind(socket_addr)?;
+    udp_socket.set_broadcast(true)?;
+    udp_socket.set_read_timeout(timeout)?;
+    let dest_socket_addr = SocketAddr::new(IpAddr::V4(broadcast), 9999);
     for _ in 0..3 {
-        socket.send_to(&req[4..req.len()], "255.255.255.255:9999")?;
+        let _send_res = udp_socket.send_to(&request[4..request.len()], dest_socket_addr);
     }
 
     let mut buf = [0_u8; 4096];
-
     let mut devices = HashMap::new();
-    while let Ok((size, addr)) = socket.recv_from(&mut buf) {
+    while let Ok((size, addr)) = udp_socket.recv_from(&mut buf) {
         let data = protocol::decrypt(&mut buf[0..size]);
         if let Ok(device_data) = serde_json::from_str::<DeviceData>(&data) {
             devices.insert(addr, device_data);
         }
     }
+    Ok(devices)
+}
 
-    Ok(devices.into_iter().collect())
+/// Discover TPLink smart devices on the local network
+///
+/// # Errors
+///
+/// Will return `Err` if there is a `io::Error` communicating with the device or
+/// a problem decoding the response.
+///
+/// # Panics
+///
+/// Will panic if protocol encryption fails.
+#[allow(clippy::needless_collect)] // needed for achieving parallelism
+pub fn with_timeout(timeout: Option<Duration>) -> Result<Vec<(SocketAddr, DeviceData)>> {
+    let request = protocol::encrypt(QUERY).unwrap();
+    let addrs = if_addrs::get_if_addrs()?;
+    thread::scope(|s| {
+        let handles = addrs
+            .into_iter()
+            .filter_map(|intf| {
+                if let Some((ip, broadcast)) = can_interface_broadcast(intf) {
+                    let request = &request;
+                    Some(s.spawn(move |_| discover_on_interface(timeout, ip, broadcast, request)))
+                } else {
+                    None
+                }
+            })
+            .collect::<Vec<_>>();
+        handles
+            .into_iter()
+            .flat_map(|join_handle| {
+                join_handle.join().ok().and_then(Result::ok).map_or_else(
+                    || HashMap::<SocketAddr, DeviceData>::new().into_iter(),
```

Comment:
:confused:

Reply:
Should we change the return of `discover_on_interface` to `Vec<(SocketAddr, DeviceData)>? It might simplify the empty iterator expression from `map_or_else`.

Reply:
That depends on @robyoung, I would use a for loop and/or disable the `pedantic` clippy warnings since they're less than great.

Reply:
I've just had a look back through the commits and I agree, it was easier to read. I'd be happy with either allowing on this function or even dropping pedantic on the whole crate. It helped me when I was writing this largely because I wasn't sure what idiomatic rust was supposed to look like.